### ISSUE TITLE
Frame an approach for CER to handle profiles of the CTDL & CEASN

### ISSUE THREAD

**Issue by stuartasutton** [CredentialEngine/CredentialRegistry#223]
Through the Navy CRADA, we are looking at specialized (profile) applications of CTDL and CEASN. Functionally, these profiles will be comprised of the two base languages as well as extensions and refinements to the base terms. Extensions and refinements of such profiles can take the following forms:

1. Refinements through creation of new sub-properties and sub-classes for base language properties and classes;
2. Addition of new properties and classes; 
3. Value constraints particular to the profiles (value vocabularies); and
4. Profile-unique cardinality requirements.

Initially, this issue asks two key questions:

1. Should CER ingest and otherwise handle data based on these profiles? and
2. If so, should it handle a full ingest (validate against) the profiles? or
3. Ingest but:
    * Ignore (discard) what it does not understand or what does not meet minimal data requirements; or
    * Ignore (discard) new properties and classes and "dumb-down" formally declared sub-classes or sub-properties (unknown to the CER) to their more general properties known to the CER (see no. 1 above under "profile additions include")?

One of the implications would be the need to handle multiple, profile-specific validation mechanism in addition to the current CE profile's validation constraints.

**Comment by siuc-nate** [comment]
Per our 1/15/2019 meeting:
- This issue was moved to the Credential Registry repository
- Suggested definition for Profiles from [https://w3c.github.io/dxwg/profiles/](https://w3c.github.io/dxwg/profiles/):

A named set of constraints on one or more identified base specifications, including the identification of any implementing subclasses of datatypes, semantic interpretations, vocabularies, options and parameters of those base specifications necessary to accomplish a particular function.

This definition includes what are often called "application profiles", "metadata application profiles", or "metadata profiles".

**Comment by cwd-mparsons** [comment]
@science @rsaksida 
The primary problem that we want to solve is to allow publishing different profiles for CER data. I think that the use of communities might fulfill this requirement. The data that is currently published to the registry is considered a registry profile, based on the classes and properties currently handled. 
 I don't recall if there are limitations to communities in this regard. As noted earlier in this thread, there may be variations in the profiles, but would be based on the same base schema. I think that one of the requirements of communities is that the same schema has to be used for all communities (at least on the same server).

**Comment by stuartasutton** [comment]
A profile of would need to accommodate:

1. Refinement of the profiled schema's existing properties and classes through creation of new subclasses and subproperties;
2. Inclusion of newly coined classes and properties tailored to the profile community's needs (coined new or borrowed from a different namespace); 
3. New constraints on value spaces (i.e., different controlled vocabularies); and
4. Other usual constraints (different cardinality, optionality, defined subset of properties and classes from the schema being profiled).

**Comment by siuc-nate** [comment]
Essentially we just need a schema that:
- enables defining other schemas
- enables defining constraints on schemas

Then the profiles could be published and maintained like any other JSON data. Once we're able to publish CTDL, CTDL-ASN, and the registry profile of each to the registry itself, and get them back out successfully, we'll know we've got it right.

I should caution though that allowing different profiles means that, unless we do something to prevent it, a given profile might not require as much data as the registry profile requires (e.g. a credential in some other profile might only require the name, which would mean that that data is incomplete if we try to load it into our system, which could lead to a number of errors).  It's also possible that a profile might (and probably will) require data that is not present in vanilla CTDL - meaning vanilla CTDL data would be incomplete (and thus lead to errors) in a system meant to consume data from that profile.

**Comment by siuc-nate** [comment]
Would there be any kind of restrictions placed on this? Once you enable profiling, you enable the addition, removal, and modification of classes and properties - effectively turning the credential registry into a **schema** registry. Is it enough that we'd only let certain organizations publish schemas/profiles, or would some other limits need to be imposed?

**Comment by stuartasutton** [comment]
@siuc-nate, just a point of clarification. A profile does not include modification of classes and properties; however, creating a subclass or subproperty of an existing class or property is not a modification.

**Comment by cwd-mparsons** [comment]
@stuartasutton Would all refinements/additions to profile schema have to be part of the 'base' CTDL schema? If not, then I don't believe the community approach would work.

**Comment by stuartasutton** [comment]
No, bullet 2 in my comment [above](https://github.com/CredentialEngine/CredentialRegistry/issues/223#issuecomment-454853674) includes the coining of new terms which should likely be in their own namespace defining the profile. We clearly would not want the CTDL to grow just because a new profile has a term that is unique to the discourse of community of practice for which it is being defined.

**Comment by stuartasutton** [comment]
There are a number of policy issues (should we/shouldn't we) that should be set aside for later with the focus here on the technical issues of accommodating such profiles.

**Comment by edgarf** [comment]
We are closing this for now. It will be re-opened once the issue resurfaces.
